controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -331,24 +333,113 @@ private ReaderGroupConfiguration getRGConfigurationFromRecord(final String scope
+                                .thenCompose(newstate -> streamMetadataStore.updateReaderGroupVersionedState(scope, readerGroup,
+                                        ReaderGroupState.ACTIVE, newstate, context, executor))
+                                .thenApply(v1 -> CreateReaderGroupStatus.Status.SUCCESS);
+                    }
+                    return Futures.failedFuture(new IllegalStateException(String.format("Error creating StateSynchronizer Stream for Reader Group %s: %s",
+                            readerGroup, createStatus.toString())));
+                })).exceptionally(ex -> {
+            log.debug("RG Stream Create failed with error: " + ex.getMessage());
+            Throwable cause = Exceptions.unwrap(ex);
+            throw new CompletionException(cause);
+        });
+    }
+
+    public CompletableFuture<CreateReaderGroupStatus.Status> createReaderGroupInternal(String scope, String rgName, ReaderGroupConfig config, final long createTimestamp) {
+        Preconditions.checkNotNull(scope, "ReaderGroup scope is null");
+        Preconditions.checkNotNull(rgName, "ReaderGroup name is null");
+        Preconditions.checkNotNull(config, "ReaderGroup config is null");
+        Preconditions.checkArgument(createTimestamp >= 0);
+        try {
+            NameUtils.validateReaderGroupName(rgName);
+        } catch (IllegalArgumentException | NullPointerException e) {
+            log.warn("Create ReaderGroup failed due to invalid name {}", rgName);
+            return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.INVALID_RG_NAME);
+        }
+
+        return RetryHelper.withRetriesAsync(() -> {
+            // 1. check if scope with this name exists...
+            return streamMetadataStore.checkScopeExists(scope)
+                    .thenCompose(exists -> {
+                        if (!exists) {
+                            return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SCOPE_NOT_FOUND);
+                        }
+                        //2. check state of the ReaderGroup
+                        return isRGCreationComplete(scope, rgName)
+                                .thenCompose(complete -> {
+                                    if (!complete) {
+                                        return streamMetadataStore.addReaderGroupToScope(scope, rgName, config.getReaderGroupId())
+                                                .thenCompose(x -> createReaderGroupTasks(scope, rgName, config, System.currentTimeMillis()))
+                                                .thenApply(y -> CreateReaderGroupStatus.Status.SUCCESS);
+                                    }
+                                    return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SUCCESS);
+                                });
+                    });
+        }, e -> Exceptions.unwrap(e) instanceof RetryableException, 10, executor);
```

Comment:
we should use RetryableException.isRetryable(e) predicate.

Reply:
Can be changed in another PR where we change this for all code using RetryableException

Reply:
please open a github issue for this. thanks